### DISCUSSION TITLE
supabase realtime authenticated channels

### DISCUSSION THREAD

**Discussion by david-plugge** [supabase/supabase#8162] [Questions] [answered]
### Update: We're currently working on Realtime Channel Authentication. You can follow the progress here: https://github.com/orgs/supabase/discussions/22484

---

I just took a look into the presence and broadcast functionality and i wonder if there is a way to limit access to specific channels. Do i have to write custom server code or is there a builtin way to solve this?

_My use case:_
In my application there are a few servers communicating with a main server. Only the main server should be accessible from the outside. I need realtime communication for process logs etc. in the browser. 
Something like socket io would be my alternative.

**Comment by w3b6x9** [comment, marked as answer]
@david-plugge there's no built-in channel auth at the moment. You'll have to do this yourself but we have discussed it internally and will prioritize accordingly.

**Reply by dungeon2567** [reply]
How do i write authentication for channels?

**Reply by jake-danton** [reply]
@w3b6x9 could you give a high level outline of what "doing it yourself" looks like? Looking at the docs, it seems all clients share the same connection strings, so no way to limit access that way. And no easy hooks to do it per broadcast event. Or read-write permissions.

It would be great if the client took a unique connection string that encoded the room, userId, and a back end-generated signature to verify that the user is who they say they are and that they have access to that room. This would solve the "google meet chat"-type cases: private access, but free for all once in.

And a per message hook that called an edge function to choose to block or modify the event would be great as well to make sure clients were not broadcasting junk messages. This could be used with or without the above. This would be great for more serious "multiplayer" features and all cases that require trusting the events.

Another option would be to have readonly channels for clients, making all events writes go through the back end first. This would make it less performant than peer-to-peer but would be great for many use cases that don't require DB writes but benefit from broadcasting updates to everyone. 

If anything like the above is possible now, that would be great!

**Reply by w3b6x9** [reply]
@JacobDanton great suggestions here! i'm going to pass this on to the Realtime team b/c they're currently working on a authorization system for realtime.

**Reply by w3b6x9** [reply]
looping in @chasers and @filipecabaco who are spearheading the authz work

**Reply by TomasHubelbauer** [reply]
@w3b6x9 sorry if this is already answered and obvious to others, but you're talking about in the future there being a behavior where after calling `supabase.auth.signInWith*`, the Realtime channels will automatically become authenticated mirroring the behavior the JS SDK already has where if the `SupabaseClient` is authenticated the calls to the REST API are and if it isn't they just use the public anon key?

Say with code like this:
```typescript
const channel = supabase
  .channel("items")
  .on(
    "postgres_changes",
    { event: "*", schema: "public", table: "items" },
    (payload) => { … }
  )
  .subscribe();
```

I wonder if this Realtime auth integration means this code would just start to work right away if the RLS for the table restricts all access to auth'd users and on the client side the `supabase.auth.signInWith*` method was called or if the solution is more general and there would be a need to adapt this code and explicitly pass auth?

**Reply by msubodh** [reply]
Hi, if this has been answered elsewhere, please let me know. However, to me this is a big security concern. I wanted to channels in a multi tenant environment. Given that the connection can be created using the public auth key, anyone savvy enough to look at the source code in a browser would just be able to get at the channels for my other clients and gain access to potentially secure and sensitive information. 

If there isnt a way to secure channels, I might have to scrap supabase as an option for this particular usecase altogether.

**Reply by filipecabaco** [reply]
We’ve prioritized this work. Hope to have something for you soon!

**Reply by Mortalife** [reply]
Has there been any progress over the last 3 months that you can share? Otherwise I think I'll just integrate partykit instead. https://docs.partykit.io/guides/authentication/

**Reply by filipecabaco** [reply]
Still no timeline as this is still in development but we're getting closer

Please do check the PRs for more details but be aware things might change

https://github.com/supabase/realtime/pulls?q=is%3Apr+is%3Aclosed

hopefully we'll be able to give more information soon but it's ongoing and currently the top priority

ps: thank you for keeping us accountable 🙏

**Reply by d-e-h-i-o** [reply]
@filipecabaco Thank you for working on this! Just one question: In the current implementation, are hard to guess channels (e.g. a uuid) secure? Or is there some way a third party could query the currently open channels to find the names?

**Reply by filipecabaco** [reply]
Still TBD as first we want to see the authz layer done but that has been discussed. we just want to first see the authorization done to then focus on building usability features.

**Reply by d-e-h-i-o** [reply]
@filipecabaco Thanks for the answer – Sorry, I think I didn't express myself clearly, I didn't mean the implementation you are working on. I meant the implementation that is currently live. So if we are using Supabase today, with the released feature set, could we open a "secure" channel by giving it a name that is impossible to guess? Or is there currently a way for a client to discover a channel by other means?

**Reply by filipecabaco** [reply]
Users cannot query for channels as we don't list them at the moment. Still I think that if a user changes their channel to `*` they would capture all of them so do test your approach against https://realtime.supabase.com/inspector

**Comment by denishacquin** [comment]
Would love to see that as well, or in the mean time if someone could indicate implementation steps it would be great!

**Comment by vinceglb** [comment]
Have there been any changes since last year? This is still a great feature to be able to integrate natively with Supabase.

**Comment by BeMoreDog** [comment]
I agree, it makes a lot of sense to add this layer of protection to avoid the attack surface that it opens at the moment (where any client can find your key and broadcast or listen to anything).
